From 265aaffe456b7fbf4b29c0feb498581deddb90e7 Mon Sep 17 00:00:00 2001
From: Dain Sundstrom <dain@apache.org>
Date: Mon, 23 Jul 2007 19:53:27 +0000
Subject: [PATCH] DBCP-225 throw an IllegalStateException if connection factory
 returns a null connection.  This avoids creating DelegatingConnections with a
 null delegate.

git-svn-id: https://svn.apache.org/repos/asf/jakarta/commons/proper/dbcp/trunk@558846 13f79535-47bb-0310-9956-ffa450edef68
---
 .../org/apache/commons/dbcp/PoolableConnectionFactory.java   | 3 +++
 .../commons/dbcp/datasources/CPDSConnectionFactory.java      | 5 +++++
 .../commons/dbcp/datasources/KeyedCPDSConnectionFactory.java | 5 +++++
 3 files changed, 13 insertions(+)

diff --git a/src/java/org/apache/commons/dbcp/PoolableConnectionFactory.java b/src/java/org/apache/commons/dbcp/PoolableConnectionFactory.java
index 6b91db5d13..2c13fd0c9e 100644
--- a/src/java/org/apache/commons/dbcp/PoolableConnectionFactory.java
+++ b/src/java/org/apache/commons/dbcp/PoolableConnectionFactory.java
@@ -292,6 +292,9 @@ public void setDefaultCatalog(String defaultCatalog) {
 
     synchronized public Object makeObject() throws Exception {
         Connection conn = _connFactory.createConnection();
+        if (conn == null) {
+            throw new IllegalStateException("Connection factory returned null from createConnection");
+        }
         if(null != _stmtPoolFactory) {
             KeyedObjectPool stmtpool = _stmtPoolFactory.createPool();
             conn = new PoolingConnection(conn,stmtpool);
diff --git a/src/java/org/apache/commons/dbcp/datasources/CPDSConnectionFactory.java b/src/java/org/apache/commons/dbcp/datasources/CPDSConnectionFactory.java
index 747fd54017..4a8259a256 100644
--- a/src/java/org/apache/commons/dbcp/datasources/CPDSConnectionFactory.java
+++ b/src/java/org/apache/commons/dbcp/datasources/CPDSConnectionFactory.java
@@ -175,6 +175,11 @@ public synchronized Object makeObject() {
             } else {
                 pc = _cpds.getPooledConnection(_username, _password);
             }
+
+            if (pc == null) {
+                throw new IllegalStateException("Connection pool data source returned null from getPooledConnection");
+            }
+
             // should we add this object as a listener or the pool.
             // consider the validateObject method in decision
             pc.addConnectionEventListener(this);
diff --git a/src/java/org/apache/commons/dbcp/datasources/KeyedCPDSConnectionFactory.java b/src/java/org/apache/commons/dbcp/datasources/KeyedCPDSConnectionFactory.java
index ce53e7baa9..e477d638c1 100644
--- a/src/java/org/apache/commons/dbcp/datasources/KeyedCPDSConnectionFactory.java
+++ b/src/java/org/apache/commons/dbcp/datasources/KeyedCPDSConnectionFactory.java
@@ -159,6 +159,11 @@ public synchronized Object makeObject(Object key) throws Exception {
         } else {
             pc = _cpds.getPooledConnection(username, password);
         }
+
+        if (pc == null) {
+            throw new IllegalStateException("Connection pool data source returned null from getPooledConnection");
+        }
+
         // should we add this object as a listener or the pool.
         // consider the validateObject method in decision
         pc.addConnectionEventListener(this);
